Allow tensorflow dataset to fetch chunks

daniel-falk opened the pull request:
As described in #1885  iterating a small dataset with `ds.tensorflow()` is very slow.

If using the tensorflow dataset on a hub dataset which consist of multiple samples with small size, the iteration becomes considerably faster if using fetch_chunks=True when getting the data. This commit adds the ability to pass that option when creating the tensorflow dataset from the hub dataset.

On an MNIST dataset the result is 25 times faster iteration.

Benchmark can be found here:
https://gist.github.com/daniel-falk/188b96013e9f0cedcf555a0a30fa177d

TF chunks: 957 frames / second
TF current: 26 frames / second
Hub chunks: 1294 frames / second
Hub per sample: 28 frames / second 

## 🚀 🚀 Pull Request

### Checklist:

- [x]  [My code follows the style guidelines of this project](https://www.notion.so/activeloop/Engineering-Guidelines-d6e502306d0e4133a8ca507516d1baab) and the [Contributing document](https://github.com/activeloopai/Hub/blob/release/2.0/CONTRIBUTING.md)
- [x]  I have commented my code, particularly in hard-to-understand areas
- [ ]  I have kept the `coverage-rate` up
- [x]  I have performed a self-review of my own code and resolved any problems
- [ ]  I have checked to ensure there aren't any other open [Pull Requests](https://github.com/activeloopai/Hub/pulls) for the same change
- [ ]  I have described and made corresponding changes to the relevant documentation
- [ ]  New and existing unit tests pass locally with my changes

daniel-falk commented:
I can also add to the benchmarking that by batching the samples I reach essentially the same throughput with tensorflow (1245 frames / second) as with the hub dataset directly (1294 frames / second):
```
batch_size = 128
iter(hub_ds.tensorflow(fetch_chunks=True).batch(batch_size)))
```

mikayelh commented:
@daniel-falk thank you so much for the contribution, this is 🔥 💣 . Looping in @tatevikh to assign reviewers.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1589,12 +1590,15 @@ def tensorflow(
         Args:
             tensors (List, Optional): Optionally provide a list of tensor names in the ordering that your training script expects. For example, if you have a dataset that has "image" and "label" tensors, if ``tensors=["image", "label"]``, your training script should expect each batch will be provided as a tuple of (image, label).
             tobytes (bool): If ``True``, samples will not be decompressed and their raw bytes will be returned instead of numpy arrays. Can also be a list of tensors, in which case those tensors alone will not be decompressed.
+            fetch_chunks: See fetch_chunks argument in hub.core.tensor.Tensor.numpy()
```
I don't think this decision should be exposed to the user, we should always fetch chunks in the context of tensorflow. What do you think @AbhinavTuli ?

daniel-falk replied:
I run the benchmark again with larger images (coco) and there was still a significant speed-up (29 fps instead of 7 fps). I do however assume that getting the first sample might be really slow if the images are large if we always use chunks, so perhaps there should be an option to disable it at least.

Coco benchmark here: https://gist.github.com/daniel-falk/e827be6abf4eb81c7083ec1732b5bb39

farizrahman4u replied:
@daniel-falk With tensorflow, we know we are going to be fetching the whole dataset (unless there are step slices) so the time to download the first sample is not relevant imo.

daniel-falk replied:
@farizrahman4u There are cases when I just want to take one or a few samples from my tensorflow dataset. One example is for visualization of a sample to debug augmentations, another example is when i need a "representable dataset" for integer quantization. In both cases I need to use the tensorflow dataset instead of the hub dataset since I want the transformations. But I agree with your point, it is no big issue if there are a few seconds of extra delay. For a 640x480x3 image, taking the first sample goes from 1 second to 5 second if using `fetch_chunks=True`.

farizrahman4u replied:
@daniel-falk interesting, in that case can you set the default for fetch chunks to True here?

daniel-falk replied:
@farizrahman4u Yes, I think that is a good option.

daniel-falk commented:
I created a simple example of how I typically use the datasets to measure the timing with and without chunked fetching, that did however result in an exception when training with the dataset with `fetch_chunks=True`... This needs to be investigated.
```
hub.util.exceptions.ReadOnlyModeError: Modification when in read-only mode is not supported!
```

Full example is here:
https://gist.github.com/daniel-falk/ec6b0dab5514b6ad54e08b666bc841b7

AbhinavTuli commented:
Hey @daniel-falk ! Thanks for reporting the issue. #1911 should fix it and we will include it in the next release.

daniel-falk commented:
Thanks @AbhinavTuli , with that fix it works good. With the large images in my "predict weather" example above, the training of the model was ~30% faster when fetching chunks.

I have changed the default value to `True` for `fetch_chunks` and updated to docstring to `deeplake` instead of `hub`.

mikayelh commented:
@daniel-falk , this is the first oss contribution to Deep Lake! Yay! :) hit you up in the community channel for some swag goodies. :)
